### ISSUE TITLE
Nested dependent: :destroy not called. Expected behavor?

### ISSUE THREAD

**Issue by bryanculver** [rails/rails#32765]
### Steps to reproduce

1. Setup nested has_many, has_many :through, has_one associations between models (see [this gist](https://gist.github.com/bryanculver/35e97d56988a00ac47560268e1e0e18e))
2. Create records to fill associations (see `sample_data.yml` in example above)
3. Update has_many collection via has_many :through association that is missing user originally in collection (example pertaining to above):
```
event = Event.find(1)
user_a = User.find(1)

event.users = [user_a] # Now missing user_b
```

### Expected behavior
Records attached via has_one association should be deleted.

### Actual behavior
Records persist but are now "orphaned"

### System configuration
**Rails version**: 5.1.5

**Ruby version**: 2.4.1p111

**Comment by dmitry** [comment]
Looks like an expected behaviour.

**Comment by LHBrave** [comment]
:+1:

**Comment by bryanculver** [comment]
I'm going to close this issue. I have reviewed the sections handling this behavior and I have resolved that ActiveRecord is doing the best it can be expected to do for two reasons:

- ActiveRecord would have to be aware of the relationships to be mindful to clean them up. For efficiency sake, ActiveRecord can just directly write the SQL queries instead of loading each object into memory. The memory overhead of creating an object for each of the records to soon be deleted is almost always wasteful. 
- The fall back catch for the stranding of these relationships would be foreign key designations in schema. However since I have laid out a polymorphic association it's obvious that you won't have that catch to throw an error that you will be stranding records. ActiveRecord may catch this ahead of time and properly cleanup but that's beyond my intention for testing.

Suffice it to say, I am comfortable calling this an edge case. I would like to thank all of the contributors for their well thought balance between simplification and convenience.
